tortoise/tortoise.go

```diff
@@ -1156,3 +1147,121 @@ func computeExpectedVoteWeight(atxdb atxDataProvider, epochWeight map[types.Epoc
 	}
 	return total, nil
 }
+
+// computeBallotWeight compute and assign ballot weight to the weights map.
+func computeBallotWeight(
+	atxdb atxDataProvider,
+	weights map[types.BallotID]*big.Float,
+	ballot *types.Ballot,
+	layerSize,
+	layersPerEpoch uint32,
+) (*big.Float, error) {
+	if ballot.EpochData != nil {
+		var (
+			total, weight uint64
+			foundWeight   bool
+		)
+		for _, atxid := range ballot.EpochData.ActiveSet {
+			atx, err := atxdb.GetAtxHeader(atxid)
+			if err != nil {
+				return nil, fmt.Errorf("atx %s in active set of %s is unknown", atxid, ballot.ID())
+			}
+			total += atx.GetWeight()
+			if atxid == ballot.AtxID {
+				weight = atx.GetWeight()
+				foundWeight = true
+			}
+		}
+		// TODO is it valid not to include your own atxid into active set?
+		if !foundWeight {
+			atx, err := atxdb.GetAtxHeader(ballot.AtxID)
+			if err != nil {
+				return nil, fmt.Errorf("atx %s in active set of %s is unknown", ballot.AtxID, ballot.ID())
+			}
+			weight = atx.GetWeight()
```

Comment:
if the ballot's ATXID not in the active set, then miner is not eligible to vote. we should return zero weight instead, even if it is found in the atx db

Reply:
should be probably validated in proposals handler